### ISSUE TITLE
Metabase not able to infer list data type from Spark Thrift Server (underlying db HDFS)

### ISSUE THREAD

**Issue by ankitsoni** [metabase/metabase#10186]
Metabase version : 0.31.3
HDFS version: 2.7.7
Spark version: 2.4.3

OS Infomation
Distributor ID:	Ubuntu
Description:	Ubuntu 19.04
Release:	19.04
Codename:	disco

Also using sparksql-deps plugin (https://github.com/metabase/sparksql-deps)

Steps to reproduce:

1 Write a paraquet file in HDFS having some list fields.
2 Start the Spark Thrift JDBC server.
3 Use beeline to create a DB (create database db_name).
4 Create a view to fetch data from the HDFS. (create view view_name AS select field1,field2....fieldn from parquet.`hdfs://HOST:PORT/path`)
4 Start the metbase and setup DB.
5 Sync the DB.
6 GO to the dashboard and select the created DB then the table.

The page shows Error:
**There was a problem with your question**

On clicking _Show error details_ we get **Table 'view_name' has no Fields associated with it.**

In logs we get the following stacktace:

06-17 10:15:29 DEBUG metabase.middleware :: GET /api/table/35/fks 200 (2 ms) (4 DB calls). Jetty threads: 8/50 (6 busy, 2 idle, 0 queued)
06-17 10:15:29 WARN metabase.query-processor :: {:status :failed,
 :class java.lang.Exception,
 :error "Table 'view_name' has no Fields associated with it.",
 :stacktrace
 ("--> query_processor.middleware.add_implicit_clauses$fn__31085$add_implicit_fields__31090$fn__31096.invoke(add_implicit_clauses.clj:81)"
  "query_processor.middleware.add_implicit_clauses$fn__31085$add_implicit_fields__31090.invoke(add_implicit_clauses.clj:64)"
  "query_processor.middleware.add_implicit_clauses$fn__31179$add_implicit_mbql_clauses__31184$fn__31190.invoke(add_implicit_clauses.clj:107)"
  "query_processor.middleware.add_implicit_clauses$fn__31179$add_implicit_mbql_clauses__31184.invoke(add_implicit_clauses.clj:105)"
  "query_processor.middleware.add_implicit_clauses$maybe_add_implicit_clauses.invokeStatic(add_implicit_clauses.clj:116)"
  "query_processor.middleware.add_implicit_clauses$maybe_add_implicit_clauses.invoke(add_implicit_clauses.clj:114)"
  "query_processor.middleware.add_row_count_and_status$add_row_count_and_status$fn__31239.invoke(add_row_count_and_status.clj:15)"
  "query_processor.middleware.driver_specific$process_query_in_context$fn__32590.invoke(driver_specific.clj:12)"
  "query_processor.middleware.resolve_driver$resolve_driver$fn__35356.invoke(resolve_driver.clj:15)"
  "query_processor.middleware.bind_effective_timezone$bind_effective_timezone$fn__31785$fn__31786.invoke(bind_effective_timezone.clj:9)"
  "util.date$call_with_effective_timezone.invokeStatic(date.clj:88)"
  "util.date$call_with_effective_timezone.invoke(date.clj:77)"
  "query_processor.middleware.bind_effective_timezone$bind_effective_timezone$fn__31785.invoke(bind_effective_timezone.clj:8)"
  "query_processor.middleware.store$initialize_store$fn__37573$fn__37574.invoke(store.clj:11)"
  "query_processor.store$do_with_new_store.invokeStatic(store.clj:34)"
  "query_processor.store$do_with_new_store.invoke(store.clj:30)"
  "query_processor.middleware.store$initialize_store$fn__37573.invoke(store.clj:10)"
  "query_processor.middleware.cache$maybe_return_cached_results$fn__32207.invoke(cache.clj:127)"
  "query_processor.middleware.catch_exceptions$catch_exceptions$fn__32255.invoke(catch_exceptions.clj:64)"
  "query_processor$process_query.invokeStatic(query_processor.clj:213)"
  "query_processor$process_query.invoke(query_processor.clj:209)"
  "query_processor$run_and_save_query_BANG_.invokeStatic(query_processor.clj:322)"
  "query_processor$run_and_save_query_BANG_.invoke(query_processor.clj:316)"
  "query_processor$fn__37738$process_query_and_save_execution_BANG___37743$fn__37744.invoke(query_processor.clj:354)"
  "query_processor$fn__37738$process_query_and_save_execution_BANG___37743.invoke(query_processor.clj:340)"
  "query_processor$fn__37762$process_query_and_save_with_max_BANG___37767$fn__37768.invoke(query_processor.clj:373)"
  "query_processor$fn__37762$process_query_and_save_with_max_BANG___37767.invoke(query_processor.clj:369)"
  "api.dataset$fn__42087$fn__42090.invoke(dataset.clj:45)"
  "api.common$fn__23152$invoke_thunk_with_keepalive__23157$fn__23158$fn__23159.invoke(common.clj:435)"),
 :query
 {:type "query",
  :query {:source-table 35},
  :parameters [],
  :constraints {:max-results 10000, :max-results-bare-rows 2000},
  :info
  {:executed-by 1,
   :context :ad-hoc,
   :card-id nil,
   :nested? false,
   :query-hash [-39, -4, -72, 92, 39, 102, -107, -49, 107, 51, -66, -104, 60, 24, -94, 55, -58, -64, -99, -1, 78, 24, 98, -75, -58, -66, 12, -33, 57, -128, -115, 28],
   :query-type "MBQL"}},
 :preprocessed nil,
 :native nil}

06-17 10:15:29 WARN metabase.query-processor :: Query failure: Table 'view_name' has no Fields associated with it.
("clojure.core$ex_info.invokeStatic(core.clj:4739)"
 "clojure.core$ex_info.invoke(core.clj:4739)"
 "--> query_processor$assert_query_status_successful.invokeStatic(query_processor.clj:285)"
 "query_processor$assert_query_status_successful.invoke(query_processor.clj:277)"
 "query_processor$run_and_save_query_BANG_.invokeStatic(query_processor.clj:323)"
 "query_processor$run_and_save_query_BANG_.invoke(query_processor.clj:316)"
 "query_processor$fn__37738$process_query_and_save_execution_BANG___37743$fn__37744.invoke(query_processor.clj:354)"
 "query_processor$fn__37738$process_query_and_save_execution_BANG___37743.invoke(query_processor.clj:340)"
 "query_processor$fn__37762$process_query_and_save_with_max_BANG___37767$fn__37768.invoke(query_processor.clj:373)"
 "query_processor$fn__37762$process_query_and_save_with_max_BANG___37767.invoke(query_processor.clj:369)"
 "api.dataset$fn__42087$fn__42090.invoke(dataset.clj:45)"
 "api.common$fn__23152$invoke_thunk_with_keepalive__23157$fn__23158$fn__23159.invoke(common.clj:435)")

06-17 10:15:29 DEBUG metabase.middleware :: POST /api/dataset 200 (133 ms) (16 DB calls). Jetty threads: 8/50 (6 busy, 2 idle, 0 queued)

**Comment by flamber** [comment]
Hi @ankitsoni 
Can you please try the latest release - 0.32.9 - which comes with Spark driver.

**Comment by jestan** [comment]
It seems metabase hive/spark driver doesn't map the data types properly, this patch works for me.
I had a table with array<string> type, also same for map<primitive,<data_type>>, and don't know how to map struct/union  types 

```clojure
(defmethod sql-jdbc.sync/database-type->base-type :hive-like [_ database-type]
  (condp re-matches (name database-type)
    #"boolean"               :type/Boolean
    #"tinyint"               :type/Integer
    #"smallint"              :type/Integer
    #"int"                   :type/Integer
    #"bigint"                :type/BigInteger
    #"float"                 :type/Float
    #"double"                :type/Float
    #"double precision"      :type/Double
    #"decimal.*"             :type/Decimal
    #"char.*"                :type/Text
    #"varchar.*"             :type/Text
    #"binary*"               :type/*
    #"date"                  :type/Date
    #"time"                  :type/Time
    #"timestamp"             :type/DateTime
    #"interval"              :type/*
    #"array.*"               :type/Array
    #"map"                   :type/Dictionary
    #".*"                    :type/*))
```

**Comment by ankitsoni** [comment]
> Hi @ankitsoni
> Can you please try the latest release - 0.32.9 - which comes with Spark driver.

Updating the version to 0.32.9 doesn't resolve the issue.
Still getting the same error.

**Comment by camsaul** [comment]
Fixed by #10207

**Comment by ankitsoni** [comment]
> Fixed by #10207

This fix by @jestan worked.
